README.md
<p align="center">
  <a href="https://design.phenixthemes.com">
    <img src="https://design.phenixthemes.com/wp-content/themes/phenix-docs/_front-end/dist/img/px-logo/phenix-design-icon.svg" alt="Phenix Design logo" width="175" height="162">
  </a>
</p>

<h2 align="center">Phenix Design System</h2>

<p align="center">
  Simple, Effective, and powerful user-interface framework for faster and easier web development.
</p>
<p align="center">
  <a href="https://design.phenixthemes.com"><strong>Explore Phenix docs »</strong></a>
</p>
<p align="center">
<a href="https://github.com/EngCode/phenix-ui/issues/new?assignees=-&labels=bug&template=bug_report.yml">Report bug</a>
<a href="https://github.com/EngCode/phenix-ui/issues/new?assignees=&labels=feature&template=feature_request.yml">Request feature</a>
<a href="https://phenixthemes.com/">Themes</a>
<a href="https://blog.phenixthemes.com/">Blog</a>
</p>

<p align="center">
  <img src="https://img.shields.io/badge/build-v0.6.5-blue.svg" alt="" />
  <img src="https://img.shields.io/github/languages/code-size/EngCode/phenix-ui.svg" alt="" />
  <img src="https://img.shields.io/github/repo-size/EngCode/phenix-ui.svg" alt="" />
  <img src="https://img.shields.io/github/issues/EngCode/phenix-ui.svg" alt="" />
  <img src="https://img.shields.io/badge/%40typescript-v3.9.10-blue.svg" alt="" />
  <img src="https://img.shields.io/badge/Sass-v7.0.1-blue.svg" alt="" />
</p>
</div>

---

### Introduction

**[Phenix DS](https://design.phenixthemes.com/about-phenix/ "About Phenix")** is lightweight user-interface framework build for simple structure, fast execution, and performance is number One concern in mind it was build for personal use and later published as open-source project for web community benefits.

---

#### 🚀 Phenix Key-Features

<div>
<ul style="list-style:none;padding:0;margin:0">
    <li>✅ CSS 100kb and 17KB Gzipped</li>
    <li>✅ JavaScript 60kb and 14.5KB Gzipped</li>
    <li>✅ Builded with Sass and TypeScript</li>
    <li>✅ Built-In Powerful JavaScript plugins</li>
    <li>✅ Build Anything with CSS Utilities</li>
    <li>✅ Flexbox Grid System</li>
    <li>✅ Masonry Grid Mode</li>
    <li>✅ RTL & LTR Fully-Support</li>
    <li>✅ CSS Variable for Design System</li>
    <li>✅ Responsive for 5 Breakpoints</li>
    <li>✅ DOM Manipulation methods</li>
    <li>✅ DOM Traversing methods</li>
    <li>✅ Javascript Media Query</li>
    <li>✅ Live Event Listener</li>
    <li>✅ DOM Javascript UI Effects</li>
    <li>✅ Responsive Multimedia</li>
    <li>✅ Javascript Time Counter</li>
    <li>✅ Dropdown Buttons and Menu's</li>
    <li>✅ View-Port Animations</li>
    <li>✴️ Design Theme Switcher</li>
    <li>✴️ API Connection Handler</li>
    <li>✅ Standard Form Controls</li>
    <li>✴️ Advanced Form Controls</li>
    <li>✅ Easy Form Validation</li>
    <li>✴️ Advanced Datatables</li>
    <li>✅ CSS Utilities to build freely</li>
    <li>✅ Responsive Navigation Menu</li>
    <li>✅ Responsive Megamenu</li>
    <li>✅ Notifications System</li>
    <li>✅ Advanced Popup Modals</li>
    <li>✅ Simple Tabs & Panels</li>
    <li>✴️ Tooltip & Popovers</li>
    <li>✅ Shape Shifter Progress Bars</li>
    <li>✴️ Advanced HTML Multimedia Players</li>
    <li>✅ Lazy-Loading for Everything</li>
    <li>✅ JavaScript Data attributes API</li>
    <li>✅ CSS Animations on Demand</li>
    <li>✅ Slider/Carousel on Demand</li>
</ul>
</div>

---

#### Changelog Update v0.7.0 {WIP}

```
01 - [new] Advanced Select
02 - [new] Phenix Repeater
03 - [new] Phenix Uploader
04 - [new] Rating Controller
05 - [new] CSS Layout Groups
06 - [update] Counter Lazyloading
07 - [update] Timer Lazyloading
08 - [update] Responsive Form Controls Sizes
09 - [Doc] Carousel Slider
10 - [update] Border Utilities
11 - [new] Clip-Path Method (https://developer.mozilla.org/en-US/docs/Web/CSS/clip-path)
```

---

#### Changelog Update v0.6.5

```
01 - [upgrade] Masonry Grid Mode
02 - [new] Collapse Javascript
03 - [upgrade] Re-Build Accordion with Collapse
04 - [new] Multi-level Dropdowns and Megamenu
05 - [new] CSS Animations on Dimand
06 - [upgrade] Add Media Lightbox for Popups.
07 - [new] Notifications and Toasts
08 - [upgrade] Typography CSS Utilites
09 - [update] Animations on Demand Demand Demand Types
10 - [new] Progress Bars
11 - [update] Carousel Slider Optimizing
```

-----------------------

#### Browser Support

| ![](https://raw.githubusercontent.com/alrra/browser-logos/master/src/firefox/firefox_48x48.png) | ![](https://raw.githubusercontent.com/alrra/browser-logos/master/src/chrome/chrome_48x48.png) | ![](https://raw.githubusercontent.com/alrra/browser-logos/master/src/safari/safari_48x48.png) | ![](https://raw.githubusercontent.com/alrra/browser-logos/master/src/opera/opera_48x48.png) | ![](https://raw.githubusercontent.com/alrra/browser-logos/master/src/edge/edge_48x48.png) | ![](https://raw.githubusercontent.com/alrra/browser-logos/master/src/archive/internet-explorer_9-11/internet-explorer_9-11_48x48.png) |
|:-----------------------------------------------------------------------------------------------:|:---------------------------------------------------------------------------------------------:|:---------------------------------------------------------------------------------------------:|:-------------------------------------------------------------------------------------------:|:-----------------------------------------------------------------------------------------:|:-------------------------------------------------------------------------------------------------------------------------------------:|
| v63+                                                                                            | v84+                                                                                          | v12+                                                                                          | v64+                                                                                        | v84+                                                                                      | ~~Dead~~                                                                                                                              |

---

#### [Documentation](https://design.phenixthemes.com)

you can visit the offical website of phenix design system and read the documentation to master the frameowork : [design.phenixthemes.com](https://design.phenixthemes.com).

---

### Where to Start

to use and understand phenix design pattern you need to get familiar with all [Utilities](https://design.phenixthemes.com/category/utilities/) class names like [spacing](https://design.phenixthemes.com/spacing/)[,](https://design.phenixthemes.com/display/) [s](https://design.phenixthemes.com/sizing/)[izing](https://design.phenixthemes.com/sizing/)[, colors](https://design.phenixthemes.com/colors/)[, backgrounds](https://design.phenixthemes.com/backgrounds/) to be able to customize and change any component used in the documentation examples and builds.

---

### Installation

phenix is available through NPM and Github you will need [Node.js](https://nodejs.org/en/download/ "Download Node JS") to install and import phenix into your project with CLI, or you can clone this project and work from it directly as work enviroment

```
====== Git CLI Clone =====
git clone https://github.com/EngCode/phenix-ui.git

====== NPM Install ======
npm install phenix-ui
```

after that you will need to install Node.JS and SASS the Ruby version for compiling the source files.

---

### SASS Modules

after installing phenix-ui from npm you can import sass modules into you project sass with `@import './node_modules/phenix-sass/*'`which will import the **“*name”** module into your project, for more information about [SASS modules click here](https://design.phenixthemes.com/category/sass-customize/ "SASS Customize") also,

---

### Typescript Modules

you can import and use the TypeScript modules into your project typescript by import Phenix from `'./node_modules/phenix-ts'`  which will import Phenix plugins for you to use it in your project,

or you can use it as JavaScript module by  import Phenix from `"./node_modules/phenix-js"` inside your script file or even with html script in the example below, for more information about how to use [P.D.S TypeScript/JavaScript click here](https://design.phenixthemes.com/category/typescript-js/ "Typescript").

---

### CDN CSS

the next piece of code will load ******[Phenix D.S](https://design.phenixthemes.com/ "Home")****** Stylesheet **“LTR”** version from npm CDN Links you can copy it and place it into your document `<head>...</head>` tag.

```
<!-- Phenix CSS LTR -->
<link href="https://cdn.jsdelivr.net/npm/phenix-ui@0.6.7/dist/css/phenix.css" rel="stylesheet" />
```

### Right-to-Left (RTL) Version

if your document language is right-to-left you can use ****[Phenix D.S](https://design.phenixthemes.com/ "Home")**** Stylesheet **“RTL”** version from npm CDN Links you can copy it and place it into your document `<head>...</head>` tag.

```
<!-- Phenix CSS RTL -->
<link href="https://cdn.jsdelivr.net/npm/phenix-ui@0.6.7/dist/css/phenix-rtl.css" rel="stylesheet" />
```

---

### CDN JavaScript

the next piece of code will load ****[Phenix D.S](https://design.phenixthemes.com/ "Home")**** [JavaScript](https://design.phenixthemes.com/category/typescript-js/ "Typescript") for [ui Effects](https://design.phenixthemes.com/ui-effects/ "UI Effects") and Interactive [Components](https://design.phenixthemes.com/category/components/ "Components") like [[popups](https://design.phenixthemes.com/popup-modals/ "Popup Modals"), [sliders](https://design.phenixthemes.com/carousel-slider/ "Carousel Slider"), [animations](https://design.phenixthemes.com/animations/ "Animations"), [tabs](https://design.phenixthemes.com/tabs-panels/ "Tabs & Panels"), [data tables](https://design.phenixthemes.com/data-tables/ "Data Tables")] from npm CDN Links you can copy it and place it before your document `</body>` closing tag.

```
<!-- Phenix JS -->
<script src="https://cdn.jsdelivr.net/npm/phenix-ui@0.6.7/dist/js/phenix.js"></script>
```

---

### Credits

this project was created in 2016 under the code-name **[Tornado UI](https://tornado.phenixthemes.com/)** until 2022 the name was changed to fit the new personal-business brand **[Phenix Themes](https://phenixthemes.com/ "https://phenixthemes.com")**.

- [Abdullah Ramadan](https://www.facebook.com/Eng.AbdallahPS) - Phenix Founder
- Naotoshi Fujita - [SplideJS]([](https://github.com/Splidejs/splide)) Creator
- Daniel Eden - [Animate.css](https://animate.style/) Creator

--------------------

### License

Copyright (c) 2016-present, [Abdullah Ramadan](https://www.facebook.com/Eng.AbdallahPS). Licensed under the terms of the [MIT License](https://opensource.org/licenses/MIT).


changelog.md
### Changelog Update v1.0.0 Final

```
Documentation Completed
the Final Release for the First Version of Phenix Design System.
```

---

### Changelog Update v0.9.0

```
01 - [new] HTML Audio Player
02 - [new] HTML Video Player
```

---

### Changelog Update v0.8.0

```
01 - [new] Phenix Data-Tables
02 - [new] Tooltip & Popovers
03 - [new] Offcanvas Components
05 - [new] Element Remover
06 - [new] Range Slider
07 - [new] Slider three.js Animations
08 - [new] Dynamic Colored Words
09 - [new] Slider Grid Addons
```

---

### Changelog Update v0.7.0

```
01 - [new] Advanced Select
02 - [new] Phenix Repeater
03 - [new] Phenix Uploader
04 - [new] Rating Controller
05 - [new] CSS Layout Groups
06 - [update] Counter Lazyloading
07 - [update] Timer Lazyloading
08 - [update] Responsive Form Controls Sizes
09 - [Doc] Carousel Slider
```

---

#### Changelog Update v0.6.5

```
01 - [upgrade] Masonry Grid Mode
02 - [new] Collapse Javascript
03 - [upgrade] Re-Build Accordion with Collapse
04 - [new] Multi-level Dropdowns and Megamenu
05 - [new] CSS Animations on Dimand
06 - [upgrade] Add Media Lightbox for Popups.
07 - [new] Notifications and Toasts
08 - [upgrade] Typography CSS Utilites
09 - [update] Animations on Demand Demand Demand Types
10 - [new] Progress Bars
11 - [update] Carousel Slider Optimizing
```

---

### Credits

this project was created in 2016 under the code-name **[Tornado UI](https://tornado.phenixthemes.com/)** until 2022 the name was changed to fit the new personal-business brand **[Phenix Themes](https://phenixthemes.com/ "https://phenixthemes.com")**.

- [Abdullah Ramadan](https://www.facebook.com/Eng.AbdallahPS) - Phenix Founder
- Naotoshi Fujita - [SplideJS](%5B%5D(https://github.com/Splidejs/splide)) Creator
- Daniel Eden - [Animate.css](https://animate.style/) Creator

### License

Copyright (c) 2016-present, [Abdullah Ramadan](https://www.facebook.com/Eng.AbdallahPS). Licensed under the terms of the [MIT License](https://opensource.org/licenses/MIT).

src/scripts/index.ts
/*!
 * Phenix.js
 * Version  : 0.6 Beta
 * License  : MIT
 * Copyright: 2022 Abdullah.Ramadan
 */

/**======> Referance By Comment <======
 * ===> 01 - Phenix Object
 * ===> 02 - D.O.M Ready
 * ===> 03 - Add Class
 * ===> 04 - Remove Class
 * ===> 05 - Toggle Class
 * ===> 06 - Get Ancestor [Parent, Parents Until]
 * ===> 07 - Get Siblings
 * ===> 08 - Get Next Siblings
 * ===> 09 - Get Previous Siblings
 * ===> 10 - CSS Styling
 * ===> 11 - Set Multiple Attributes
 * ===> 12 - Insert Elements
 * ===> 13 - Event Handler
 * ===> 14 - Resources Lazy-Loader
 * ===> 15 - Media Query Method
 * ===> 16 - Service Worker
 * ===> 17 - Define Info Grapers
 * ===> 18 - Define UI Effects
 * ===> 19 - Define Other Features
 * ===> 20 - Include Features
 * ===> 21 - Integration
 * ===> 22 - Phenix Selecting Method
 * ===> 23 - Your Custom Script [JS]
*/

/*====> Phenix Object <====*/
export class PhenixElements extends Array<HTMLElement | Object | 'object'> {
    /*====> D.O.M Ready <====*/
    ready(callback:any) {
        //====> Check if its Ready <====//
        if (document.readyState == 'complete') callback();

        //====> Wait for It to be Ready <====//
        else document.addEventListener('DOMContentLoaded', callback);

        //====> Return Phenix Elements <====//
        return this;
    }

    /*====> Add Class <====*/
    addClass(className:string) {
        //====> Add Class for Each Element <====//
        this.forEach((element:HTMLElement) => element.classList.add(className));

        //====> Return Phenix Elements <====//
        return this;
    }

    /*====> Remove Class <====*/
    removeClass(className:string) {
        //====> Add Class for Each Element <====//
        this.forEach((element:HTMLElement) => element.classList.remove(className));

        //====> Return Phenix Elements <====//
        return this;
    }

    /*====> Toggle Class <====*/
    toggleClass(className:string) {
        //====> Toggle Class for Each Element <====//
        this.forEach((element:HTMLElement) => element.classList.toggle(className));

        //====> Return Phenix Elements <====//
        return this;
    }

    /*====> Get Ancestor <====*/
    ancestor(target?) {
        //====> Define Ancestor Arrays <====//
        let ancestors = [];

        //====> Loop Through Phenix Elements <====//
        this.forEach((element:HTMLElement) => {
            //====> Define The First Parent <====//
            let parent:any = element.parentNode;

            //====> if has a Target <===//
            if (target && !element.matches('html')) {
                //====> Loop Over The Ancestors <====//
                while (parent && !parent.matches('html')) {
                    //====> When the Target has been Found Return it <====//
                    if (parent.matches(target)) { ancestors.push(parent); break; }
                    //====> Otherwise get the Next Ancestor <====//
                    else parent = parent.parentNode;
                }
            }

            //====> Otherwise Return the Parent <====//
            else ancestors.push(parent);
        });

        //====> Return Ancestors <====//
        if (ancestors.length > 1) return ancestors;
        else if (ancestors.length === 1) return ancestors[0];
    }

    /*====> Get Siblings <====*/
    siblings(target?) {
        //====> Siblings Define <====//
        let siblings = [];

        //====> Loop Through Phenix Elements <====//
        this.forEach((element:any) => {
            //====> Get This Element Parent Childs <====//
            let childs = element.parentNode.children;

            //====> if has No target [Return All Sibling] <====//
            if (!target) Array.from(childs).forEach((child:HTMLElement) => child !== element ? siblings.push(child) : null);

            //====> Otherwise Return the matched targets <====//
            else Array.from(childs).forEach((child:HTMLElement) => child !== element && child.matches(target) ? siblings.push(child) : null);
        });

        //====> Return Siblings <====//
        return siblings.length > 0 ? siblings : null;
    }

    /*====> Get Next Siblings <====*/
    next(target?, all_target?) {
        //====> Sibling Define <====//
        let siblings = [];
        
        //====> Loop Through Phenix Elements <====//
        this.forEach((element:any) => {
            //====> Define Next Unit <====//
            let nextUnit = element.nextElementSibling;
    
            //====> if No target [Return the Direct Next Sibling] <====//
            if (!target) siblings.push(nextUnit);
    
            //====> if Target is Detected <====//
            if (target) {
                //====> if All Next is the Target <====//
                if (target.includes('all', 0)) {
                    //====> Loop Through All Next Siblings <====//
                    while (nextUnit) {
                        //====> if All has Target & Matches this Unit <====//
                        if (all_target && nextUnit.matches(all_target)) siblings.push(nextUnit);
    
                        //====> if All has no Target get this Unit <====//
                        if (!all_target) siblings.push(nextUnit);
    
                        //====> get the next one <====//
                        nextUnit = nextUnit.nextElementSibling;
                    }
                }
    
                //====> Otherwise get the Next Matched Target [Next Until] <====//
                else {
                    //====> While there is a Next Unit <====//
                    while (nextUnit) {
                        //====> Check for the Target & Return it <====//
                        if (nextUnit.matches(target)) { siblings.push(nextUnit); break; } 
                        
                        //====> Otherwise Get the Next Unit <====//
                        else nextUnit = nextUnit.nextElementSibling;
                    }
                }
            }
        });

        //====> Return Siblings <====//
        if (siblings.length > 1) return siblings;
        else if (siblings.length === 1) return siblings[0];
    }

    /*====> Get Previous Siblings <====*/
    prev(target?, all_target?) {
        //====> Sibling Define <====//
        let siblings = [];
        
        //====> Loop Through Phenix Elements <====//
        this.forEach((element:any) => {
            //====> Define Previous Unit <====//
            let prevUnit = element.previousElementSibling;
    
            //====> if No target [Return the Direct Previous Sibling] <====//
            if (!target) siblings.push(prevUnit);
    
            //====> if Target is Detected <====//
            if (target) {
                //====> if All Previous is the Target <====//
                if (target.includes('all', 0)) {
                    //====> Loop Through All Previous Siblings <====//
                    while (prevUnit) {
                        //====> if All has Target & Matches this Unit <====//
                        if (all_target && prevUnit.matches(all_target)) siblings.push(prevUnit);
    
                        //====> if All has no Target get this Unit <====//
                        if (!all_target) siblings.push(prevUnit);
    
                        //====> get the previous one <====//
                        prevUnit = prevUnit.previousElementSibling;
                    }
                }
    
                //====> Otherwise get the Previous Matched Target [Previous Until] <====//
                else {
                    //====> While there is a Previous Unit <====//
                    while (prevUnit) {
                        //====> Check for the Target & Return it <====//
                        if (prevUnit.matches(target)) {
                            siblings.push(prevUnit);
                            break;
                        } else {
                            prevUnit = prevUnit.previousElementSibling;
                        }
                    }
                }
            }
        });
        
        //====> Return Siblings <====//
        if (siblings.length > 1) return siblings;
        else if (siblings.length === 1) return siblings[0];
    }

    /*====> Get Childrens <====*/
    child(target?:string) {
        //====> Childrens Define <====//
        let childs = [];

        //====> Loop Through Phenix Elements <====//
        this.forEach((element:any) => {
            //====> All Childrens <====//
            let all_childs = element.children;
    
            //====> if No target [Return All Direct Childrens] <====//
            if (!target) childs = all_childs;
    
            //====> if Target is Matched Return it <====//
            if (target) all_childs.forEach(element => element.matches(target) ? childs.push(element) : '');
        });

        //====> Return Childrens <====//
        if (childs.length > 1) return childs;
        else if (childs.length === 1) return childs[0];
    }

    /*====> CSS Styling <====*/
    css(style:object, clearInline?) {
        //====> if inline-style Clear is Activated <====//
        if (clearInline) this.forEach((element:HTMLElement) => (element.removeAttribute('style')));

        //====> for Each CSS Property <====//
        for (const [key, value] of Object.entries(style)) {
            //====> Convert Property[String] To Object Name[Key] <====//
            // let property = key.replace(/(-[a-z])/, prop => prop.replace("-", "").toUpperCase());

            //====> Style Elements <====//
            this.forEach((element:HTMLElement) => {
                element.style.setProperty(key, value);
                // element.style[property] = value
            });
        }

        //====> Return Phenix Elements <====//
        return this
    }

    /*====> Set Multiple Attributes <====*/
    setAttributes(attributes:{}) {
        //====> Loop Through Phenix Elements <====//
        this.forEach((element:HTMLElement) => {
            //====> Loop Through the Attributes <====//
            Object.keys(attributes).forEach(attribute => {
                //====> Set Attributes One By One <====//
                element.setAttribute(attribute, attributes[attribute]);
            });
        });

        //====> Return Phenix Elements <====//
        return this;
    }

    /*====> Insert Elements <====*/
    insert(position?:string, elementsPackage?:any) {
        //====> Define the Package <====//
        let new_package = [];

        //====> Loop Through Phenix Elements <====//
        this.forEach((element:HTMLElement) => {
            //===> String Element Converter to [Node] <===//
            if (typeof(elementsPackage) === 'string')
                elementsPackage = document.createRange().createContextualFragment(elementsPackage);

            //====> Insert After The Element <====//
            if (position === 'after') {
                element.parentNode.insertBefore(elementsPackage, element.nextSibling);
                //====> Include the New Element to the Package <====//
                new_package.push(element.nextElementSibling);
            }

            //====> Insert Before The Element <====//
            else if (position === 'before') {
                element.parentNode.insertBefore(elementsPackage, element);
                //====> Include the New Element to the Package <====//
                new_package.push(element.previousElementSibling);
            }

            //====> Insert Inside The Element at First <====//
            else if (position == 'prepend') {
                element.prepend(elementsPackage);
                //====> Include the New Element to the Package <====//
                new_package.push(element.firstChild);
            }

            //====> Insert Inside The Element at Last <====//
            else if (position == 'append') {
                element.appendChild(elementsPackage);
                //====> Include the New Element to the Package <====//
                new_package.push(element.lastChild);
            }
        });

        //====> Return the New Elements Package <====//
        if (new_package.length > 1) return new_package;
        else if (new_package.length === 1) return new_package[0];
    }

    /*====> Event Handler <====*/
    on(event, callback, live?, timer?) {
        //====> Attache a Normal Event <====//
        if (!live) {
            //====> Event for Each Element <====//
            this.forEach((element:HTMLElement) => {
                element.addEventListener(event, callback);
            });

            //====> Return Phenix Elements <====//
            return this;

        //====> Live Event <====//
        } else if (live && this.length > 0) {
            //====> Create Time Loop & Attach Event <====//
            let timerLoop = setInterval(() => this.forEach((element:HTMLElement) => element.addEventListener(event, callback)), timer | 1000);

            //====> Return the Elements & Time Loop <====//
            return {
                elements: this,
                timeLoop: timerLoop
            };
        }
    }

    /*====> Resources Lazy-Loader <====*/
    lazyLoading() {
        //====> Element Data <====//
        let spiner = "data:image/svg+xml;base64,PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0idXRmLTgiPz4KPHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHhtbG5zOnhsaW5rPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5L3hsaW5rIiBzdHlsZT0ibWFyZ2luOiBhdXRvOyBiYWNrZ3JvdW5kOiByZ2JhKDAsIDAsIDAsIDApIG5vbmUgcmVwZWF0IHNjcm9sbCAwJSAwJTsgZGlzcGxheTogYmxvY2s7IHNoYXBlLXJlbmRlcmluZzogYXV0bzsiIHdpZHRoPSIyMDBweCIgaGVpZ2h0PSIyMDBweCIgdmlld0JveD0iMCAwIDEwMCAxMDAiIHByZXNlcnZlQXNwZWN0UmF0aW89InhNaWRZTWlkIj4KPGNpcmNsZSBjeD0iNTAiIGN5PSI1MCIgZmlsbD0ibm9uZSIgc3Ryb2tlPSIjZGNkY2RjIiBzdHJva2Utd2lkdGg9IjMiIHI9IjE4IiBzdHJva2UtZGFzaGFycmF5PSI4NC44MjMwMDE2NDY5MjQ0MSAzMC4yNzQzMzM4ODIzMDgxMzgiPgogIDxhbmltYXRlVHJhbnNmb3JtIGF0dHJpYnV0ZU5hbWU9InRyYW5zZm9ybSIgdHlwZT0icm90YXRlIiByZXBlYXRDb3VudD0iaW5kZWZpbml0ZSIgZHVyPSIxcyIgdmFsdWVzPSIwIDUwIDUwOzM2MCA1MCA1MCIga2V5VGltZXM9IjA7MSI+PC9hbmltYXRlVHJhbnNmb3JtPgo8L2NpcmNsZT4KPCEtLSBbbGRpb10gZ2VuZXJhdGVkIGJ5IGh0dHBzOi8vbG9hZGluZy5pby8gLS0+PC9zdmc+";
        
        //====> Loop Through Media Elements <====//
        Phenix('img, video, audio, iframe').forEach((element:HTMLElement) => {
            //====> Set Loading Mode <====//
            if (!Phenix(element).inView()) {
                //====> Get Data <====//
                let playable = element.matches('video' || 'audio'),
                    source = !playable ? element.getAttribute('src') : null,
                    preloaded = element.getAttribute('preload' || 'loading');

                //===> for [images, iframe] <===//
                if (element.matches('img'||'iframe')) {
                    // element.setAttribute('data-lazyload', source);
                    // element.setAttribute('src', spiner);
                    element.classList.add('px-loading')
                }

                //===> for [video, audio] <===//
                else if (playable && !preloaded) element.setAttribute('preload', 'none');
                element.setAttribute('loading', 'lazy');

                //====> Keep Watching Element While Scrolling <====//
                document.addEventListener('scroll', event => {
                    //====> if its in view-point set the original source <====//
                    if (element.matches('img'||'iframe') && Phenix(element).inView() && !element.matches('.px-loaded')) {
                        // element.setAttribute('src', source);
                        element.classList.remove('px-loading');
                        element.classList.add('px-loaded');
                    };
                });
            }

            //====> First View [workout] <====//
            else if (Phenix(element).inView() && element.matches('img') && element.getAttribute('data-lazyload')) {
                element.setAttribute('src', element.getAttribute('data-lazyload'));
                element.classList.add('px-loaded');
            }
        });

        //====> Other Elements Types <====//
        Phenix('.px-counter').forEach((counter:HTMLElement) => {
            if(!counter.classList.contains('counting')) document.addEventListener('scroll', event => {
                if(Phenix(counter).inView() && !counter.classList.contains('counting')) Phenix(counter).counter();
            });
        });

        //====> Return Phenix Elements <====//
        return this;
    }

    /*====> Media Query <====*/
    mediaQuery(breakpoint:any, fn_callback?:any, mobile_first?:boolean) {
        /*==== Breakpoints Points ====*/
        let xsmall_end = 485, /*===> Max xSmall breakpoint Width */
            small_start= 580, /*===> Min Small breakpoint Width */
            small_end = 705,  /*===> Max Small breakpoint Width */
            /*==== Medium Points ====*/
            medium_start = 775,  /*===> Min Medium breakpoint Width */
            medium_end = 1105,   /*===> Max Medium breakpoint Width */
            /*==== Large Points ====*/
            large_start = 1205, /*===> Min Large breakpoint Width */
            large_end = 1370,   /*===> Max Large breakpoint Width */
            /*==== xLarge Points ====*/
            xlarge_start = 1405, /*===> Min xLarge breakpoint Width */
            xlarge_end = 2565;   /*===> Max xLarge breakpoint Width */

        /*==== Activator ====*/
        const createQuery = () => {
            //==== Get Current Screen Width ====//
            let current_vw = Phenix(document).viewport('width');

            /*==== Extra Small ====*/
            if (mobile_first && breakpoint == 'xsmall') {
                if (current_vw >= 0) {
                    fn_callback();
                }
            } else if (breakpoint == 'xsmall') {
                if (current_vw <= xsmall_end) {
                    fn_callback();
                }
            }

            /*==== Small ====*/
            if (mobile_first && breakpoint == 'small')  {
                if (current_vw >= small_start) {
                    fn_callback();
                }
            } else if (breakpoint == 'small')  {
                if (current_vw <= small_end) {
                    fn_callback();
                }
            }

            /*==== Medium ====*/
            if (mobile_first && breakpoint == 'medium') {
                if (current_vw >= medium_start) {
                    fn_callback();
                }
            } else if (breakpoint == 'medium') {
                if (current_vw <= medium_end) {
                    fn_callback();
                }
            }

            /*==== Large ====*/
            if (mobile_first && breakpoint == 'large')  {
                if (current_vw >= large_start) {
                    fn_callback();
                }
            } else if (breakpoint == 'large')  {
                if (current_vw <= large_end) {
                    fn_callback();
                }
            }

            /*==== Extra Large ====*/
            if (mobile_first && breakpoint == 'xlarge') {
                if (current_vw >= xlarge_start) {
                    fn_callback();
                }
            } else if (breakpoint == 'xlarge')  {
                if (current_vw <= xlarge_end) {
                    fn_callback();
                }
            }

            /*==== Custom Media ====*/
            if (mobile_first && typeof(breakpoint) == 'number')   {
                if (current_vw >= breakpoint) {
                    fn_callback();
                }
            } else if (typeof(breakpoint) == 'number')  {
                if (current_vw <= breakpoint) {
                    fn_callback();
                }
            }
        };

        //===> Active <===//
        createQuery();
        //===> Live Check <===//
        window.addEventListener('resize', resizing => {
            createQuery();
        });

        //====> Return Phenix Elements <====//
        return this;
    }

    /*====> Define Info Grapers <====*/
    height; getCSS; direction;
    inView; viewport; copyrights;

    /*====> Define UI Effects <====*/
    slideUp; slideDown; slideToggle;
    fadeOut; fadeIn; fadeToggle;

    /*====> Define Other Features <====*/
    counter; multimedia; timer;
    sticky; smothScroll; scrollSpy;
    design; themes; animations;

    /*====> Define Components <====*/
    dropdown; tabs; collapse;
    popup; lightbox; menu; datatable;
    validation; uploader; progress;
    select; repeater; rating;
    slider; utilities; notifications;
}

/*====> Phenix Selecting Method <====*/
const Phenix = (selector?:any) => {
    /*====> Get Elements <====*/
    if (typeof(selector) === 'string') {
        //====> Select as Phenix Elements <====//
        let selected = document.querySelectorAll(selector);

        //====> Create Elements Query <====//
        if (selector.length > 0) return new PhenixElements(...selected);
    /*====> if its Elements Passed it <====*/
    } else if (selector !== null && typeof(selector) !== 'undefined' || 'number') {
        //====> if Not Array Make it one <====//
        if (!Array.isArray(selector) || typeof(selector) !== 'object') selector = [selector];

        //====> and Create Elements Query <====//
        return new PhenixElements(...selector);
    }
}

//====> Export Phenix <====//
export default Phenix;

/*====> Import Features <====*/
import './features/get-info';   //==> Info Grapers
import './features/effects';    //==> UI Effects
import './features/counter';    //==> Animated Counter
import './features/animations'; //==> View-port Animations
import './features/validation'; //==> Form Validation
import './features/collapse';   //==> Collapse Toggle
import './features/notifications'; //==> Notifications

/*====> Import Components <====*/
import './components/menu';      //==> Menus
import './components/media';     //==> Media Setter
import './components/timer';     //==> Time Counter
import './components/dropdown';  //==> Dropddown
import './components/tabs';      //==> Tabs System
import './components/popup';     //==> Popups
import './components/progress';  //==> Progress
import './components/uploader';  //==> File Uploader
import './components/datatable'; //==> Data-Tables

/*====> Integration <====*/
import './integration/slider';    //==> Splide.js Slider
import './integration/utilities'; //==> Phenix Utilities
import './integration/wordpress'; //==> Wordpress Integration

/*====> Custom Script <====*/
import './custom-scripts';

//====> Export Global Phenix <====//
module.exports = Phenix;